### PR TITLE
Corey/data split

### PR DESCRIPTION
**Pull request by Ckoelewyn**
This PR covers task **Create a new train/val from train (possibly need stratification)**. 

Use the `pickle_maker.ipynb` notebook for creating pickles in your GDrive
Use the `--pickle` argument with `train.py` in `colab_train.ipynb` to specify which pickle you want to train on.

Features
 * df_pickler module with functions for creating pickle files from dataframes and reading data frames from pickle files
   * Creates training, validation, and testing datasets (training and validation come from COCO's training, testing comes from COCO's testing)
 * Notebook file `pickle_maker.ipynb` for creating pickle files in your google drive
 * new arg `--pickle` specifies which pickle you want to consume
    * If pickle flag not used train.py uses `hourglass.load_and_filter_annotations()` to generate training and validation dataframes
    * If the pickle doesn't exist, an error is output and train.py uses `hourglass.load_and_filter_annotations()` to generate training and validation dataframes
  * new script `anno_dl.sh` for downloading only annotation zips.

Potential issues
  * naming scheme for model folders won't reflect which pickle this came from
  * stratification is not currently implemented, could result in training and validation sets having unequal representation of different parameters.

### CODE REVIEW

File: hourglass.py [hunk 1]
```diff
@@ -53,10 +54,13 @@ def load_and_filter_annotations(self, path_to_train_anns,path_to_val_anns, subse
         print(f"Train/Val dfs contains {len(train_df)}/{len(val_df)} anns")
         return train_df.reset_index(), val_df.reset_index()
 
-    def _start_train(self, batch_size, model_base_dir, epochs, initial_epoch, model_subdir, current_time, subset):
+    def _start_train(self, batch_size, model_base_dir, epochs, initial_epoch, model_subdir, current_time, subset, pickle_name):
         self._compile_model()
-
-        train_df, val_df = self.load_and_filter_annotations(DEFAULT_TRAIN_ANNOT_PATH, DEFAULT_VAL_ANNOT_PATH, subset)
+        # insert logic here for pickled dataframes
+        if pickle_name != None:
```
Comment (julianrocha) [inline hourglass.py]: Use `is not` instead of `!=`
"Comparisons to singletons like None should always be done with is or is not, never the equality operators."
[PEP 8 Style Guide](https://www.python.org/dev/peps/pep-0008/#programming-recommendations)

File: df_pickler.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    df = df.loc[df['bbox_area'] > drop_small_bbox]
+    df_TrV = df.loc[df['source'] == 0] #Data from the training set for splitting into train/val
+    df_Tst = df.loc[df['source'] == 1] #Data from the val set for testing
+    t_size = scale * TV_split
+    v_size = scale * (1-TV_split)
+    df_Tr, df_V = train_test_split(df_TrV, train_size=t_size, test_size=v_size) # We can stratify here if wanted as well
+    df_Tr.to_pickle(os.path.join(pickle_folder, f'Train_{pickle_name}.pkl'))
+    df_V.to_pickle(os.path.join(pickle_folder, f'Val_{pickle_name}.pkl'))
+    df_Tst.to_pickle(os.path.join(pickle_folder, f'Test_{pickle_name}.pkl'))
+    return
```
Comment (julianrocha) [inline df_pickler.py:51]: These are hard-coded filtering rules like we have in `load_and_filter_annotations`. I think one of the benefits we get with pickling is the ability to filter/partition the dataset as we want, and can easily make new pickles for different types of splits/filters. For this reason, I would argue that the filtering rules applied here should be moved to the responsibility of the user of `pickle_maker.ipynb`. So the user can create a `coco_df`, and filter what they want into 3 dfs, and then could pass all 3 to `make_pickle`. This is very much a question of requirements, I'm suggesting this logic is moved to the notebook, but I'm curious what others think as this currently offers the benefit of documenting our filters/partitions in code rather than hoping that whoever created the pickle filtered correctly.

File: df_pickler.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    return df_Tr.reset_index(), df_V.reset_index(), df_Tst.reset_index()
+
+def pickle_namer(is_crowd, drop_empty, scale, TV_split, drop_small_bbox):
+    filename = f''
+    filename += f'C-{is_crowd}_'
+    filename += f'DE-{drop_empty}_'
+    filename += f'S-{int(scale*100)}_'
+    filename += f'TVs-{int(TV_split*100)}_'
+    filename += f'DSBB-{drop_small_bbox}'
+    return filename
```
Comment (julianrocha) [inline df_pickler.py:68]: Should we decide to move filtering logic to the notebook level, then we would likely need to pass custom filenames to `make_pickle` instead of having a `pickle_namer`

File: train.py [hunk 4]
```diff
@@ -83,6 +85,9 @@ def process_args():
     argparser.add_argument('--resume-subdir',
                         default=None,
                         help='Subdirectory containing architecture json and weights')
+    argparser.add_argument('--pickle',
+                        default=None,
+                        help='Name of folder with pickled dataframes')
```
Comment (julianrocha) [inline train.py:94]: It might be nice to merge some basic pickles into the repository so that everyone has the same set of `train/val/test` to work with. Obviously people could create new ones but a basic set to get started is nice to quickly clone a new repo and start training without needing to download anything else. How big are these files btw? (not sure if we wanna avoid checking big files into the repo)

File: hourglass.py [hunk 5]
```diff
@@ -53,10 +54,13 @@ def load_and_filter_annotations(self, path_to_train_anns,path_to_val_anns, subse
...
+            train_df, val_df, test_df = get_pickle(pickle_name)
```
Comment (julianrocha) [inline hourglass.py:64]: Since `test_df` likely won't be used in this method, could change the unpacking to: `train_df, val_df, _`

File: scripts/anno_dl.sh [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# Adapted from coco_dl.sh
+# Removed the image datasets, they are not required for creating dataframe objects
```
Comment (julianrocha) [inline scripts/anno_dl.sh]: noice addition

Comment (robertklee) [inline scripts/anno_dl.sh]: I'm wondering if this script can be merged with the existing `coco_dl.sh` script. It seems all we need to add is another command line argument after dataset path. (From first glance I don't see anything new in this script except for the removal of the image zip download)

**Comment by robertklee** [comment]
I think you accidentally deleted `submodules/HeatMap`

File: pickle_maker.ipynb [hunk 7]
```diff
@@ -0,0 +1 @@
...loss function matches across training sessions"],"execution_count":null,"outputs":[{"output_type":"stream","text":["WARNING:tensorflow:From /tensorflow-1.15.2/python3.7/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n","\n","WARNING:tensorflow:From /tensorflow-1.15.2/python3.7/keras/callbacks/tensorboard_v1.py:200: The name tf.summary.merge_all is deprecated. Please use tf.compat.v1.summary.merge_all instead.\n","\n","WARNING:tensorflow:From /tensorflow-1.15.2/python3.7/keras/callbacks/tensorboard_v1.py:203: The name tf.summary.FileWriter is deprecated. Please use tf.compat.v1.summary.FileWriter instead.\n","\n","2021-03-23 08:17:48.371388: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7\n","2021-03-23 08:17:49.616987: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10\n","WARNING:tensorflow:From /tensorflow-1.15.2/python3.7/keras/callbacks/tensorboard_v1.py:343: The name tf.Summary is deprecated. Please use tf.compat.v1.Summary instead.\n","\n"],"name":"stdout"}]}]}
```
Comment (julianrocha) [inline pickle_maker.ipynb]: <p>Duplicated</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/robertklee/COCO-Human-Pose/pull/15/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (julianrocha) [inline pickle_maker.ipynb]: <p>The <code>make_pickle()</code>  looks nice to use. If we decide to move logic to the notebook, then we can create <code>coco_df</code>  here , filter it and split it into <code>train,val,test</code>  dfs and pass them along with the filename to <code>make_pickle()</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/robertklee/COCO-Human-Pose/pull/15/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (julianrocha) [inline pickle_maker.ipynb]: <p>This train cell should probs be removed. The <code>--pickle</code>  flag for training should be moved to <code>colab-train.ipynb</code>  so that people know how to use it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/robertklee/COCO-Human-Pose/pull/15/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

File: .gitignore [hunk 8]
```diff
@@ -132,4 +132,7 @@ dmypy.json
 .pyre/
 
 # VS Code Settings
-.vscode/*
\ No newline at end of file
+.vscode/*
+
+#Pickles and pickle directories
+Pickles/*
```
Comment (robertklee) [inline .gitignore]: nit: end with newline character for better diffs

**Review by julianrocha — COMMENTED** [review]
Looks great. Left some comments, we should clarify if the suggestion I made is what we want. Also, this PR removes the HeatMap Submodule which we should not do (since we still need this submodule)

File: hourglass.py (hunk 5 above)
Comment (robertklee) [reply hourglass.py:64]: we may need some way to load test (which for us might be the val2017 set) separately than the others, since we'll be running evaluation on the test set. We'll have to decide how many entries to set aside for validation and testing.

File: pickle_maker.ipynb (hunk 7 above)
Comment (robertklee) [inline pickle_maker.ipynb]: I think for clearer diffs, could you remove the outputs and throw the JSON in a JSON prettifier?

**Comment by robertklee** [comment]
Overall it looks like a great addition for some much needed data splitting options for us. I haven't fully reviewed it (I'm working on the assignment) but I've left a few comments on things I noticed. Thanks for knocking this off our backlog.

File: train.py (hunk 4 above)
Comment (Ckoelewyn) [reply train.py:94]: Full set of Train/Val/Test with no scaling is 151MB

Comment (robertklee) [reply train.py:94]: We could deploy them via GitHub Package Releases. Up to 1 GB per file, and I believe it also generates a link we can use in our scripts. What was the downside to doing the splitting in-memory? Was it so we have consistent train/val splits on the training set (i.e. no chance of a different split across training sessions)? Could we just save a csv of annotation IDs, and populate the data frames using those IDs?

Comment (Ckoelewyn) [reply train.py:94]: Consistency between the sets. If we get ejected from colab deep into a training set, having the splits saved would allow training to resume with identical training/validation/testing as before. If we wanted to train for multiple days across the same dataset we no longer need to worry about the 24hr timeout. 

Saving it as a Csv is doable, but It’s [much slower than pickle](https://towardsdatascience.com/the-best-format-to-save-pandas-data-414dca023e0d?gi=c431087feed5)

I’ll look into GitHub package releases tomorrow.

Comment (robertklee) [reply train.py:94]: Sounds good, thanks for the clarification. Fully agree with your analysis. Then I think GitHub package releases might be the best option for us. We would be limited to 2 GB per 'release' which sounds like it's plenty of breathing room for us. We'd need to confirm that the package URL stays constant.

File: scripts/anno_dl.sh (hunk 6 above)
Comment (Ckoelewyn) [reply scripts/anno_dl.sh]: Per our discussion this afternoon I've wrapped the image downloading portion in `coco_dl.sh`. Passing `no_image` as the second argument prevents the script from downloading the image dataset.

**Review by robertklee — APPROVED** [review]
LGTM!
